validator/client/beacon-api/duties.go

```diff
@@ -5,21 +5,228 @@ import (
+		}
+
+		syncDutiesMapping[types.ValidatorIndex(validatorIndex)] = true
+	}
+
+	// Mapping from the {committeeIndex, slot} to each of the committee's validator indices
+	committeeMapping := make(map[committeeIndexSlotPair][]types.ValidatorIndex)
+	for _, committee := range committees {
+		committeeIndex, err := strconv.ParseUint(committee.Index, 10, 64)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse committee index `%s`", committee.Index)
+		}
+
+		slot, err := strconv.ParseUint(committee.Slot, 10, 64)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse slot `%s`", committee.Slot)
+		}
+
+		validatorIndices := make([]types.ValidatorIndex, len(committee.Validators))
+		for index, validatorIndexString := range committee.Validators {
+			validatorIndex, err := strconv.ParseUint(validatorIndexString, 10, 64)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to parse committee validator index `%s`", validatorIndexString)
+			}
+			validatorIndices[index] = types.ValidatorIndex(validatorIndex)
+		}
+
+		key := committeeIndexSlotPair{
+			committeeIndex: types.CommitteeIndex(committeeIndex),
+			slot:           types.Slot(slot),
+		}
+		committeeMapping[key] = validatorIndices
+	}
+
+	duties := make([]*ethpb.DutiesResponse_Duty, len(multipleValidatorStatus.Statuses))
+	for index, validatorStatus := range multipleValidatorStatus.Statuses {
+		validatorIndex := multipleValidatorStatus.Indices[index]
+		pubkey := multipleValidatorStatus.PublicKeys[index]
+
+		var attesterSlot types.Slot
+		var committeeIndex types.CommitteeIndex
+		var committeeValidatorIndices []types.ValidatorIndex
+
+		if committeeMappingKey, ok := attesterDutiesMapping[validatorIndex]; ok {
+			committeeIndex = committeeMappingKey.committeeIndex
+			attesterSlot = committeeMappingKey.slot
+
+			if committeeValidatorIndices, ok = committeeMapping[committeeMappingKey]; !ok {
+				return nil, errors.Errorf("failed to find validators for committee index `%d` and slot `%d`", committeeIndex, attesterSlot)
+			}
+		}
+
+		duties[index] = &ethpb.DutiesResponse_Duty{
+			Committee:       committeeValidatorIndices,
+			CommitteeIndex:  committeeIndex,
+			AttesterSlot:    attesterSlot,
+			ProposerSlots:   proposerDutySlots[types.ValidatorIndex(validatorIndex)],
+			PublicKey:       pubkey,
+			Status:          validatorStatus.Status,
+			ValidatorIndex:  types.ValidatorIndex(validatorIndex),
+			IsSyncCommittee: syncDutiesMapping[types.ValidatorIndex(validatorIndex)],
+		}
+	}
+
+	return duties, nil
+}
+
+// GetCommittees retrieves the committees for the given epoch
+func (c beaconApiDutiesProvider) GetCommittees(ctx context.Context, epoch types.Epoch) ([]*apimiddleware.CommitteeJson, error) {
+	committeeParams := url.Values{}
+	committeeParams.Add("epoch", strconv.FormatUint(uint64(epoch), 10))
+	committeesRequest := buildURL("/eth/v1/beacon/states/head/committees", committeeParams)
```

Comment:
I want to double check if using the head in this case is correct, the grpc one read a little differently.

Reply:
The way I understand it from the spec, the `epoch` query parameter decides which epoch to query the committees for. The `head` committees will only be fetches if `epoch` isn't specified.

epoch | Fetch committees for the given epoch. If not present then the committees for the epoch of the state will be obtained.
-- | --

Reply:
got it this makes sense.